### ISSUE TITLE
replace call to functions from the string module in the python module

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4999]
|<img alt="nueces's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| nueces reported|
|-|-|
|Trac ID|trac#4999|
|Type|enhancement|
|Created|2011-03-27 17:29:25Z|
|Branch|https://github.com/twisted/twisted/tree/str-methods-python-4999-2|

To make the port to python 3k more easy, I'm replacing all the calls to functions from the string module, for their equivalent from the str type object or calling directly to the built-in method of the string variable.

Attachments:

* [python.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/299/2996abdb7066821d2b008c83fa572c8059609867/e5ab992ec2a98da24fb5f3dddf42bc7710587e5c.patch) (5500 bytes) - added by nueces on 2011-03-27 17:29:37Z - patch

<details><summary>Searchable metadata</summary>

```
trac-id__4999 4999
type__enhancement enhancement
reporter__nueces nueces
priority__normal normal
milestone__Python_3_x Python-3.x
branch__branches_str_methods_python_4999_2 branches/str-methods-python-4999-2
branch_author__nueces__thijs nueces, thijs
status__closed closed
resolution__fixed fixed
component__core core
keywords__py3k py3k
time__1301246965000000 1301246965000000
changetime__1344785318000000 1344785318000000
version__None None
owner__thijs thijs

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_2"></a>|@lvh commented|
|-|-|

(In [31458]) Branching to 'str-methods-python-4999'

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_3"></a>|@lvh set owner to @lvh<br>@lvh set status to `assigned`|
|-|-|

Patch looks clean, waiting for buildbot to come back clean:

http://twistedmatrix.com/trac/raw-attachment/ticket/4999/python.patch

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_4"></a>|@lvh commented|
|-|-|

Whoops! That was obviously the wrong link. I meant this:

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/str-methods-python-4999

... which looks clean except for an unrelated IOCP failure.

Added to my merge queue.

Thanks as always for your patch!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

This patch modifies untested code.

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_6"></a>|@lvh set owner to nueces<br>@lvh set status to `new`|
|-|-|

What JP said. If you don't feel like adding a test, please unassign yourself (== assign to the empty item at the top of the list), by the way :)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_9"></a>|@thijstriemstra commented|
|-|-|

(In [35057]) Branching to 'str-methods-python-4999-2'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_10"></a>|@thijstriemstra commented|
|-|-|

(In [35059]) replace string methods in t.python, refs [#4999](https://github.com/twisted/twisted/issues/4999)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_11"></a>|@thijstriemstra commented|
|-|-|

Replaced the string methos and added some new tests for the code that was changed but not covered. [Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/str-methods-python-4999-2).

**Comment by twisted-trac** [comment]
|[<img alt="vperic's avatar" src="https://avatars.githubusercontent.com/u/688281?s=50" width="50" height="50">](https://github.com/vperic)<a name="note_12"></a>|@vperic set owner to @thijstriemstra|
|-|-|

Hmm, this overlaps with my patch at [#5830](https://github.com/twisted/twisted/issues/5830) (make test_text.py pass in py3). As far as that module is concerne, the changes are the same (though I had some others, unrelated to the string module) but the tests are different. In general, I think your tests are a bit better - the problem is in test_dict - the ordering of elements in a dictionary is not cannonical so I believe it *will* fail on some architecture combination. In my patch, I tested just a single element dict, which will always work but provides less coverage (though it did cover all the changed lines so it was technically correct). Also, in the LineTests class, you might want to consider adding some assertFalse's, though I don't think it's really necessary.

Another note, (and I've been bitten by this myself), apparently Trac separates keywords with spaces, not commas, so you should use "py3k review" instead of "py3k, review". I'm also not sure what's the point of the keyword, especially when we have the milestone, so I haven't even been using it lately.

Anyway, please fix test_dict and then I'll check if all changed lines are covered (though I think they are). Consider this an official review. :)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_13"></a>|@thijstriemstra commented|
|-|-|

(In [35080]) address review comments. refs [#4999](https://github.com/twisted/twisted/issues/4999)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_14"></a>|@thijstriemstra removed owner|
|-|-|

Thanks for the review. It's unfortunate the tickets do slightly overlap but the part that was changed for [#5830](https://github.com/twisted/twisted/issues/5830) related to the string method was designated for this ticket (and maybe should've been fixed first, which were about to do :) Landing this ticket will only make the other ticket easier to review and fix.

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/str-methods-python-4999-2).

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_15"></a>|@itamarst set owner to @thijstriemstra|
|-|-|

1.Drop the changes to twisted.python.hook, they're unrelated to this branch.
 1. LineTest needs a docstring.
 1. In docstrings of test methods, rather than saying "X supports Y", you should explain what X's behavior is. What is stringyString supposed to do to tuples?

Other than that it looks good, so once above are fixed and buildbot is happy it should be merged.

**Comment by twisted-trac** [comment]
|[<img alt="vperic's avatar" src="https://avatars.githubusercontent.com/u/688281?s=50" width="50" height="50">](https://github.com/vperic)<a name="note_16"></a>|@vperic commented|
|-|-|

I think you misunderstood my comment about test_dict - any dict with more than 1 key-value pair (and it now has two) can be a problem. I'm referring to this particular line from the documentation:

"CPython implementation detail: Keys and values are listed in an arbitrary order which is non-random, varies across Python implementations, and depends on the dictionary’s history of insertions and deletions."

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_17"></a>|@itamarst commented|
|-|-|

Ah, yes. You shouldn't assume any particular order of dictionary items, as test_dict currently does. So that'd need to be fixed.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_18"></a>|@thijstriemstra commented|
|-|-|

(In [35147]) address review comments, refs [#4999](https://github.com/twisted/twisted/issues/4999)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_19"></a>|@thijstriemstra set status to `assigned`|
|-|-|

Replying to [itamar](#note_17):
> Ah, yes. You shouldn't assume any particular order of dictionary items, as test_dict currently does. So that'd need to be fixed.

Fixed.

Replying to [itamar](#note_15):
>  1. Drop the changes to twisted.python.hook, they're unrelated to this branch.

I reverted the coding standard stuff and left the string import change in there

>  1. LineTest needs a docstring.

Fixed.

>  1. In docstrings of test methods, rather than saying "X supports Y", you should explain what X's behavior is. What is stringyString supposed to do to tuples?
>

Fixed.

> Other than that it looks good, so once above are fixed and buildbot is happy it should be merged.

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/str-methods-python-4999-2).

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_20"></a>|@thijstriemstra set status to `closed`|
|-|-|

(In [35152]) Merge str-methods-python-4999-2: Replaced calls to functions from the string module in `twisted.python`.

Author: nueces, thijs
Reviewer: lvh, exarkun, vperic, itamar
Fixes: [#4999](https://github.com/twisted/twisted/issues/4999)
